csrc/fusion_segmenter.cpp

```diff
@@ -1489,6 +1490,23 @@ void convertInputRfactorsToRoots(Fusion* fusion) {
     NVF_ERROR(new_root_domain.size() == tv->domain()->contiguity().size());
     auto new_td = IrBuilder::create<TensorDomain>(
         new_root_domain, tv->domain()->contiguity());
+
+    if (tv->domain()->hasAllocation()) {
+      // we need to replay the new root domain following the old rfactor domain
```

Comment:
Could you add tests that actually require this replay? The current one only reorders the domains.

Reply:
Turns out testing this is tricky, since we don't have `convertInputRfactorsToRoots` exposed.

So we only hit this function when we go through automatic scheduler, which triggers segmentation. But automatic scheduler do not support complex transforms on allocation domain.

The reason is that, `setAllocationDomain` has a check that requires allocation domain "consistent" with leaf domain (read `nvfuser::ir_utils::validateDomainEquivalence`). So in order to set a merge or something like that on the allocation domain, I need to apply the same transformation on leaf.
But we are not supposed to hand a `transformed` fusion to automatic scheduler.

Here's the issue when I try to run this example.
```
  auto tv0 = TensorViewBuilder().ndims(3).shape({-1, -1, -1}).build();
  fusion->addInput(tv0);
  auto s0 = IrBuilder::create<Val>(5, DataType::Float);
  auto tv1 = add(tv0, s0); 
  fusion->addOutput(tv1);

  tv0->merge(0, 1);

  std::vector<IterDomain*> alloc_domain = {tv0->axis(2)};
  IterDomain* merged_id = 
      IterDomainBuilder(fusion->zeroVal(), mul(tv0->axis(0)->extent(), tv0->axis(1)->extent()))
          .parallel_type(tv0->axis(0)->getParallelType())
          .iter_type(tv0->axis(0)->getIterType())
          .build();
  IrBuilder::create<Merge>(tv0->axis(0)->container(), merged_id, tv0->axis(0), tv0->axis(1));
  alloc_domain.push_back(merged_id);

  tv0->setAllocationDomain(alloc_domain, {false, true});
```

```
C++ exception with description "derived_domain_ == frontier_ INTERNAL ASSERT FAILED at "/opt/pytorch/nvfuser/csrc/ir/utils.cpp":877, please report a bug with repro script to NVFuser at https://github.com/NVIDIA/Fuser/issues. Invalid derived domain. Initial domain: iS2{i2}, iS3{( i0 * i1 )}. Derived domain: iS0{i0}, iS1{i1}, iS2{i2}
```

If we really want to test this code path with transformation, I guess we can expose this function directly. But that feels a bit too much. tagging @naoyam for a second opinion.

Reply:
Is there a typo here?: `std::vector<IterDomain*> alloc_domain = {tv0->axis(2)};`. Since `tv0` is merged to a 2D tensor, `tv0->axis(2)` should not work, right?

Reply:
sorry I copied the wrong code.... I was trying to manually schedule it after running into the error.
Then I realized that manual scheduling won't hit the same code path, unless I expose the function.

Reply:
If you look at this
```
  std::vector<IterDomain*> alloc_domain = {tv0->axis(2)};
  IterDomain* merged_id = 
      IterDomainBuilder(fusion->zeroVal(), mul(tv0->axis(0)->extent(), tv0->axis(1)->extent()))
          .parallel_type(tv0->axis(0)->getParallelType())
          .iter_type(tv0->axis(0)->getIterType())
          .build();
  IrBuilder::create<Merge>(tv0->axis(0)->container(), merged_id, tv0->axis(0), tv0->axis(1));
  alloc_domain.push_back(merged_id);
```

You can see that I was manually merging the two iterdomain.

Reply:
I'm not entirely sure what you're trying to do here, but the error is because of the manual merge, which doesn't update the leaf domain. The check in `setAllocationDomain` is to make sure the allocation domain sits between the root and leaf domains, and now the code is trying to set an allocation domain that's outside of the path.

Conceptually, this should be completely fine since the new allocation domain is equivalent to the leaf domain. However, I don't think the current indexing works with that since it basically assumes the allocation domain consists of IterDomains that are ascendants of leaf IterDomains. Fundamentally, it should be possible to traverse from leaf IterDomains to whatever allocation domains, but that's not how it works at this moment.

For this PR, can we just set up a tensor with an allocation domain that sits between root and rfactor domains? Isn't there a test doing that already?

Reply:
both my attempts at getting the cpp test failed... got into new codegen issue. Opened #1047 #1048 to track those.

Reply:
😅

Reply:
Thanks for trying it out. Looks like there are still many things we need to do in the  scheduler, as expected. 

For the meantime, please add an assertion that the allocation domain is just an reordered rfactor domain. Or we could expose this function as public so that we could test it without going through the scheduler/segmenter. Either seems fine to me.

Reply:
I'll add an assertion for now. since that's easier to remove later when we try to fix other utils in the scheduler.